builtin/credential/approle/path_role.go

```diff
@@ -0,0 +1,1544 @@
+		lock := b.secretIDLock(secretIDHMAC)
+		lock.RLock()
+
+		result := secretIDStorageEntry{}
+		if entry, err := req.Storage.Get(entryIndex); err != nil {
+			lock.RUnlock()
+			return nil, err
+		} else if entry == nil {
+			lock.RUnlock()
+			return nil, fmt.Errorf("storage entry for SecretID is present but no content found at the index")
+		} else if err := entry.DecodeJSON(&result); err != nil {
+			lock.RUnlock()
+			return nil, err
+		}
+		listItems = append(listItems, result.SecretIDAccessor)
+		lock.RUnlock()
+	}
+
+	return logical.ListResponse(listItems), nil
+}
+
+// setRoleEntry grabs a write lock and stores the options on an role into the storage.
+// Also creates a reverse index from the role's RoleID to the role itself.
+func (b *backend) setRoleEntry(s logical.Storage, roleName string, role *roleStorageEntry) error {
+	b.roleLock.Lock()
+	defer b.roleLock.Unlock()
+
+	// Create a storage entry for the role
+	entry, err := logical.StorageEntryJSON("role/"+strings.ToLower(roleName), role)
+	if err != nil {
+		return err
+	}
+	if entry == nil {
+		return fmt.Errorf("failed to create storage entry for role %s", roleName)
+	}
+	if err = s.Put(entry); err != nil {
+		return err
+	}
+
+	// Create a storage entry for reverse mroleing of RoleID to role.
+	// Note that secondary index is created when the roleLock is held.
+	return b.setRoleIDEntry(s, role.RoleID, &roleIDStorageEntry{
+		Type: "role",
+		Name: roleName,
+	})
+}
+
+// roleEntry grabs the read lock and fetches the options of an role from the storage
+func (b *backend) roleEntry(s logical.Storage, roleName string) (*roleStorageEntry, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("missing role_name")
+	}
+
+	var result roleStorageEntry
+
+	b.roleLock.RLock()
+	defer b.roleLock.RUnlock()
+
+	if entry, err := s.Get("role/" + strings.ToLower(roleName)); err != nil {
+		return nil, err
+	} else if entry == nil {
+		return nil, nil
+	} else if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// pathRoleCreateUpdate registers a new role with the backend or updates the options
+// of an existing role
+func (b *backend) pathRoleCreateUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	// Check if the role already exists
+	role, err := b.roleEntry(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new entry object if this is a CreateOperation
+	if role == nil && req.Operation == logical.CreateOperation {
+		hmacKey, err := uuid.GenerateUUID()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create role_id: %s\n", err)
+		}
+		role = &roleStorageEntry{
+			HMACKey: hmacKey,
+		}
+	} else if role == nil {
+		return nil, fmt.Errorf("Role entry not found when the requested operation is to update it")
```

Comment:
Please have it start lowercase (general Go error rules), but it's also verbose -- "role entry not found during update operation" should suffice :-)